internal/hook/item_hook_handler.go

```diff
@@ -76,6 +79,79 @@ type ItemHookHandler interface {
 	) error
 }
 
+// ItemRestoreHookHandler invokes restore hooks for an item
+type ItemRestoreHookHandler interface {
+	// HandleRestoreHooks runs the restore hooks for an item.
+	// If the item is a pod hooks are chosen to be run as follows:
+	// If the pod has the appropriate annotations specifying the hook action, then hooks from the annotation are run
+	// Otherwise, the supplied ResourceRestoreHooks are applied.
+	HandleRestoreHooks(
+		log logrus.FieldLogger,
+		groupResource schema.GroupResource,
+		obj runtime.Unstructured,
+		rh []ResourceRestoreHook,
+	) (runtime.Unstructured, error)
+}
+
+// InitContainerRestoreHookHandler is the restore hook handler to add init containers to restored pods.
+type InitContainerRestoreHookHandler struct{}
+
+func (i *InitContainerRestoreHookHandler) HandleRestoreHooks(
+	log logrus.FieldLogger,
+	groupResource schema.GroupResource,
+	obj runtime.Unstructured,
+	resrouceRestoreHooks []ResourceRestoreHook,
+) (runtime.Unstructured, error) {
+
+	// We only support hooks on pods right now
+	if groupResource != kuberesource.Pods {
+		return nil, nil
+	}
+
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get a metadata accessor")
+	}
+	pod := new(corev1api.Pod)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.UnstructuredContent(), pod); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	initContainers := []corev1api.Container{}
+	if len(pod.Spec.InitContainers) > 0 && pod.Spec.InitContainers[0].Name == restic.InitContainer {
+		initContainers = append(initContainers, pod.Spec.InitContainers[0])
+		pod.Spec.InitContainers = pod.Spec.InitContainers[1:]
+	}
+
+	hooksFromAnnotations := getInitRestoreHookFromAnnotation(metadata.GetAnnotations())
+	if hooksFromAnnotations != nil {
+		initContainers = append(initContainers, hooksFromAnnotations.InitContainers...)
+	} else {
+		// pod did not have the annotations appropriate for restore hooks
+		// running applicable ResourceRestoreHooks supplied.
+		namespace := metadata.GetNamespace()
+		labels := labels.Set(metadata.GetLabels())
+
+		for _, rh := range resrouceRestoreHooks {
+			if !rh.Selector.applicableTo(groupResource, namespace, labels) {
+				continue
+			}
+			for _, hook := range rh.RestoreHooks {
```

Comment:
This raises a question for me - how are we enforcing order on these containers? Are annotations guaranteed to be ordered? Should we sort them before doing this?

Reply:
I am not sure I understand the question.
If we find restore hooks from the pod annotation, then we run only those restore hook and don't use the ones from the restore spec.
We look/ use the restore hooks from the spec only if the pod doesn't have the restore hook annotation.
Multiple init containers can be specified using restore spec and not annotations.
The ordering of the restore hooks in the restore spec is expected to be preserved.

I added a comment about this at at https://github.com/vmware-tanzu/velero/pull/2787/files#diff-fc409aa9f287e68f0586d526ee9a122dR87-R90